Add explicit disclosure example (StockExchange) (#25)
digital-asset/ex-java-bindings: StockExchange/README.rst

> Explicit Disclosure with Java Bindings Example

Reviewer: ```suggestion
Example of Explicit Disclosure with Java Bindings
```

----------------------------------------------

::

  Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.

>   SPDX-License-Identifier: Apache-2.0.0

Reviewer: This license is wrong, but let's not change it right now

> This project exemplifies usage of `Explicit Contract Disclosure <https://docs.daml.com/app-dev/explicit-contract-disclosure.html>`_

Reviewer: ```suggestion
This project demonstrates the usage of `Explicit Contract Disclosure <https://docs.daml.com/app-dev/explicit-contract-disclosure.html>`_
```

in Daml client applications built with the `Java Binding library <https://docs.daml.com/app-dev/bindings-java/index.html>`_.

In this example, four parties, each hosted on their own participant (see the topology configuration in `canton_ledger.conf <canton_ledger.conf>`_), are involved in a simplified trade.
Each party interacts with the Canton ledger via a standalone Java application. The example interaction flow is modelled as follows:

- Party **Bank** (see `Bank <src/main/java/examples/stockexchange/parties/Bank.java>`_) issues ``IOU`` as units of cash to the **Buyer**
- Party **StockExchange** (see `StockExchange <src/main/java/examples/stockexchange/parties/StockExchange.java>`_) issues ``Stock`` as on-ledger asset to the **Seller** party.
  Additionally, it issues price ticks for the stock as ``PriceQuotation``. Since **StockExchange** is the sole stakeholder of the ``PriceQuotation``,
  it discloses the contract for usage as reference data in commands requiring it as an input.
- Then, party **Seller** (see `Seller <src/main/java/examples/stockexchange/parties/Seller.java>`_) owns a unit of ``Stock`` issued by the **StockExchange**.

>   **Seller** creates a ``Offer`` contract on-ledger that can be accepted by any interested party (see ``Offer_Accept`` in the Daml model).

Reviewer: ```suggestion
  **Seller** creates an ``Offer`` contract on-ledger that can be accepted by any interested party (see ``Offer_Accept`` in the Daml model).
```

  Similarly to the **StockExchange**, the **Seller** discloses its ``Stock`` and ``Offer`` contracts off-ledger
  for interested parties.
- **Buyer** (see `Buyer <src/main/java/examples/stockexchange/parties/Buyer.java>`_) owns an amount of ``IOU`` issued by **Bank**.
  **Buyer** wants to exchange with the **Seller** and accepts its ``Offer`` on-ledger at the correct ``IOU`` market value in exchange of **Seller** s ``Stock``.

>   **Buyer** uses in its command submission that exercises ``Offer_Accept`` the contracts previously disclosed by the **Seller** and **StockExchange**.

Reviewer: ```suggestion
  In the command submission that exercises ``Offer_Accept``, the **Buyer** uses contracts previously disclosed by the **Seller** and **StockExchange**.
```

> **Note**: For exemplification purposes, the disclosed contracts mentioned above are shared via files in this project

Reviewer: ```suggestion
**Note**: For illustration, the disclosed contracts in this project are shared via files.
```

> (see `Common.shareDisclosedContract <src/main/java/examples/stockexchange/Common.java>`_).

Reviewer 2: @tudor-da : what's the reason for this decision? Why not use the alternative of sharing them via a HTTP server run by the stock exchange, and thereby make it match more closely the reference architecture: https://docs.google.com/presentation/d/1X7XShLONfqnZfHfquQ9qix84m59z7b6vOTi6dIIERfY/edit#slide=id.g285d43863e3_0_73

Reviewer: Tudor has already gone above and beyond to document the explicit disclosure. He extended the java-bindings component himself and then implemented an example using it. Normally it would be in the remit of the ETX team to do that. Adding a http server and all that comes with it (like inventing a protocol) has too low ROI. It would not change much to the demonstration of the ED capabilities. After all, it's about out-of-band data exchange irrespective of the means.

Author: I considered how far to go on making the look&feel real, but made a couple of trade-offs for the sake of keeping this just an example wrt time investment. If you feel strongly about this, maybe open an issue against this repo so it can be picked up at some point

Reviewer 2: I agree with you guys that adding an HTTP server would be overkill. I guess what I was looking for is a comment for readers as to what would typically be used in production for explicit disclosure, i.e., a HTTP server.

> The Daml model for the involved templates mentioned above is located in `daml/StockExchange.daml <daml/StockExchange.daml>`_`.

Reviewer: ```suggestion
The Daml model for the templates involved is located in `daml/StockExchange.daml <daml/StockExchange.daml>`_`.
```

For a better understanding of the explicit disclosure concept and off-ledger data sharing, refer to the
`Explicit Contract Disclosure <https://docs.daml.com/app-dev/explicit-contract-disclosure.html>`_ documentation
where this example's flow is also presented in more detail.

Running the example
===================

> #. If you do not have it already, download and unzip `Canton open-source <https://github.com/digital-asset/daml/releases/download/v2.8.0-snapshot.20231109.2/canton-open-source-2.8.0-snapshot.20231109.11490.0.vd02500a6.tar.gz>`_ into a location of your choice.

Reviewer: ```suggestion
#. If you do not have it already, download and unzip `Canton open-source <https://github.com/digital-asset/daml/releases/download/v2.8.0-snapshot.20231109.2/canton-open-source-2.8.0-snapshot.20231109.11490.0.vd02500a6.tar.gz>`_  or a later version into a location of your choice.
```

Reviewer: I would use the 2.8.0, it will come out soon and will be better than this snapshot
